ompi/mca/pml/ob1/custommatch/pml_ob1_custom_match_fuzzy512-byte.h

```diff
@@ -0,0 +1,594 @@
+
+#ifndef PML_OB1_CUSTOM_MATCH_FUZZY512_BYTE_H
+#define PML_OB1_CUSTOM_MATCH_FUZZY512_BYTE_H
+
+#include <immintrin.h>
+
+#include "../pml_ob1_recvreq.h"
+#include "../pml_ob1_recvfrag.h"
+
+typedef struct custom_match_prq_node
+{
+    __m512i keys;
+    __m512i mask;
+    struct custom_match_prq_node* next;
+    int start, end;
+    void* value[64];
+} custom_match_prq_node;
+
+typedef struct custom_match_prq
+{
+    custom_match_prq_node* head;
+    custom_match_prq_node* tail;
+    custom_match_prq_node* pool;
+    int size;
+} custom_match_prq;
+
+static inline int custom_match_prq_cancel(custom_match_prq* list, void* req)
+{
+#if CUSTOM_MATCH_DEBUG
+    printf("custom_match_prq_cancel - list: %x req: %x\n", list, req);
+#endif
+    __mmask64 result = 0;
+    custom_match_prq_node* prev = 0;
+    custom_match_prq_node* elem = list->head;
+    int i;
+    while(elem)
+    {
+        for(i = elem->start; i <= elem->end; i++)
+        {
+            if(elem->value[i] == req)
+            {
+#if CUSTOM_MATCH_DEBUG
+                printf("Canceled!");// %x %x %x\n", req, req->req_tag, req->req_peer);
+#endif
+                void* payload = elem->value[i];
+                ((int8_t*)(&(elem->keys)))[i] = ~0;
+                ((int8_t*)(&(elem->mask)))[i] = ~0;
+                elem->value[i] = 0;
+                if(i == elem->start || i == elem->end)
+                {
+                    while((elem->start <= elem->end) && (!(elem->value[elem->start]))) elem->start++;
+                    while((elem->start <= elem->end) && (!(elem->value[elem->end])))   elem->end--;
+                    if(elem->start > elem->end)
+                    {
+                        if(prev)
+                        {
+                            prev->next = elem->next;
+                        }
+                        else
+                        {
+                            list->head = elem->next;
+                        }
+                        if(!elem->next)
+                        {
+                            list->tail = prev;
+                        }
+                        elem->next = list->pool;
+                        list->pool = elem;
+                    }
+                }
+                list->size--;
+                return 1;
+            }
+        }
+        prev = elem;
+        elem = elem->next;
+    }
+    return 0;
+}
+
+static inline void* custom_match_prq_find_verify(custom_match_prq* list, int tag, int peer)
+{
+#if CUSTOM_MATCH_DEBUG
+    printf("custom_match_prq_find_verify list: %x tag: %x peer: %x\n", list, tag, peer);
+#endif
+    __mmask64 result = 0;
+    custom_match_prq_node* elem = list->head;
+    int i;
+    int8_t key = peer ^ tag;
+    __m512i search = _mm512_set1_epi8(key);
+    while(elem)
+    {
+        result = _mm512_cmpeq_epi8_mask(_mm512_and_epi32(elem->keys, elem->mask), _mm512_and_epi32(search, elem->mask));
+        if(result)
+        {
+            for(i = elem->start; i <= elem->end; i++)
```

Comment:
I would think that looping around the set bits in result will be faster as it saves few branches.